crunchers/cruncher.h

```diff
@@ -136,6 +138,63 @@ template <typename IMPL, size_t BUFFER_SIZE = 1024 * 1024>
 using StreamCruncher =
     current::ss::StreamSubscriber<GenericCruncherImpl<IMPL, BUFFER_SIZE>, typename IMPL::event_t>;
 
+// An extension of the `GenerucCruncherImpl` wrapper above,
+// which also generates "ticks" at specified time intervals.
+// The IMPL has the same requirements except it must also implement `OnTick()` method.
+template <typename IMPL, size_t BUFFER_SIZE>
+struct GenericTickCruncherImpl : public GenericCruncherImpl<IMPL, BUFFER_SIZE> {
+  using base_t = GenericCruncherImpl<IMPL, BUFFER_SIZE>;
+  using event_t = typename base_t::event_t;
+  using EntryResponse = current::ss::EntryResponse;
+
+  class TickMessage final : public base_t::Message {
+   public:
+    TickMessage(const std::chrono::microseconds& timestamp) : timestamp_(timestamp) {}
+    void Handle(IMPL& cruncher) override { cruncher.OnTick(timestamp_); }
+
+   private:
+    const std::chrono::microseconds timestamp_;
+  };
+
+  template <typename... ARGS>
+  GenericTickCruncherImpl(uint16_t port,
+                          const std::string& route,
+                          const std::chrono::microseconds tick_interval,
+                          ARGS&&... args)
+      : base_t(port, route, std::forward<ARGS>(args)...),
+        tick_interval_(tick_interval),
+        last_event_us_(std::chrono::microseconds(0)) {}
+  virtual ~GenericTickCruncherImpl() = default;
+
+  EntryResponse operator()(const event_t& event, idxts_t current, idxts_t last) {
+    GenerateTickEvents(current.us);
+    return base_t::operator()(event, current, last);
+  }
+
+  EntryResponse operator()(event_t&& event, idxts_t current, idxts_t last) {
+    GenerateTickEvents(current.us);
+    return base_t::operator()(std::move(event), current, last);
+  }
+
+ private:
+  void GenerateTickEvents(std::chrono::microseconds us) {
+    const uint64_t start_marker = last_event_us_.count() / tick_interval_.count();
+    const uint64_t end_marker = us.count() / tick_interval_.count();
+    for (uint64_t i = start_marker; i < end_marker; ++i) {
+      mmq_.Publish(std::make_unique<TickMessage>(std::chrono::microseconds((i + 1) * tick_interval_.count())));
```

Comment:
Why not a loop from `std::chrono::microseconds(((last_event_us_.count() / tick_interval_.count()) + 1) * tick_interval_.count())` to `<= us.count()` and `+= tick_interval_count_` ?